src/source-pcap-file.c

```diff
@@ -101,6 +147,351 @@ TmEcode DecodePcapFile(ThreadVars *, Packet *, void *, PacketQueue *, PacketQueu
+}
+
+TmEcode PcapDirectoryPopulateBuffer(
+        PcapFileDirectoryVars *ptv,
+        time_t newer_than,
+        time_t older_than,
+        time_t *first_new_file_time,
+        MemBuffer **directory_content
+) {
+    struct dirent * dir = NULL;
+
+    MemBuffer *temp_directory_content = NULL;
+
+    time_t fnt = *first_new_file_time;
+
+    while ((dir = readdir(ptv->directory)) != NULL)
+    {
+        if (dir->d_type != DT_REG)
+        {
+            continue;
+        }
+
+        char pathbuff[PATH_MAX] = {0};
+
+        int written = 0;
+
+        written = snprintf(pathbuff, PATH_MAX, "%s/%s", ptv->filename, dir->d_name);
+
+        if (written > 0 && written < PATH_MAX)
+        {
+            struct timespec temp_time;
+
+            if (PcapDirectoryGetModifiedTime(pathbuff, &temp_time) == 0)
+            {
+                SCLogDebug("File %s time (%lu > %lu < %lu)", pathbuff, newer_than, temp_time.tv_sec, older_than);
+
+                // Skip files outside of our time range
+                if (temp_time.tv_sec < newer_than) {
+                    SCLogDebug("Skipping old file %s", pathbuff);
+                    continue;
+                }
+                else if (temp_time.tv_sec > older_than) {
+                    SCLogDebug("Skipping new file %s", pathbuff);
+                    if (temp_time.tv_sec < fnt || fnt == 0)
+                    {
+                        *first_new_file_time = temp_time.tv_sec;
+                    }
+                    continue;
+                }
+            }
+
+            if (temp_directory_content == NULL)
+            {
+                temp_directory_content = MemBufferCreateNew(sizeof(char*) * 32);
+                if (!temp_directory_content)
+                {
+                    SCLogError(SC_ERR_MEM_ALLOC, "Failed to create buffer");
+
+                    SCReturnInt(TM_ECODE_FAILED);
+                }
+            }
+            else if (MEMBUFFER_SIZE(temp_directory_content) - MEMBUFFER_OFFSET(temp_directory_content) < sizeof(char*))
+            {
+                // Double size
+                if (MemBufferExpand(&temp_directory_content, MEMBUFFER_SIZE(temp_directory_content)) != 0)
+                {
+                    FreeDirectoryMemBuffer(temp_directory_content);
+
+                    SCLogError(SC_ERR_MEM_ALLOC, "Failed to expand buffer");
+
+                    SCReturnInt(TM_ECODE_FAILED);
+                }
+            }
+
+            char * mem = SCMalloc(sizeof(char *) * (written + 1));
+            if (!mem)
+            {
+                FreeDirectoryMemBuffer(temp_directory_content);
+
+                SCLogError(SC_ERR_MEM_ALLOC, "Failed to copy file name");
+
+                SCReturnInt(TM_ECODE_FAILED);
+            }
+            memcpy(mem, pathbuff, written + 1);
+
+            SCLogDebug("Found \"%s\"", mem);
+
+            *((char**)(MEMBUFFER_BUFFER(temp_directory_content) + MEMBUFFER_OFFSET(temp_directory_content))) = mem;
```

Comment:
this is definitely not the MemBuffer's intended use. It seems a bit over-complicated too. Why not use a simple list/TAILQ?